rfd/0191-workload-id-config-ux.md

```diff
@@ -0,0 +1,1454 @@
+thousands of CI workflows can share the same Bot and WorkloadIdentity.
+Where possible, we can mitigate the need for growing numbers of WorkloadIdentity
+resources by adding more advanced templating and authorization rule
+functionality.
+
+We should bear in mind the following potential performance improvements:
+
+- Tailored cache with label-based indexing to improve the performance of 
+  resolving label matchers to specific WorkloadIdentity resources.
+- Edge cache in `tbot` that can be enabled to return recently-issued identities
+  without hitting the Auth Service. This will involve some elements of the
+  authorization logic being executed at the edge for the cache.
+- Decoupling the Workload Identity CA and issuance process from the Auth
+  Service.
+
+OpenTelemetry tracing should be included across the credential issuance process
+to highlight performance hot-spots.
+
+### Deprecation
+
+For now, the old RBAC and RPCs for Workload Identity will be left in place as
+they provide effectively an "administrative" way of generating workload 
+identity credentials.
+
+When the new mechanism has been available for 2 major versions, we should
+revisit the deprecation of the legacy RPCs and RBAC.
+
+### Implementation Pathway
+
+This work is well-suited to being broken down into a number of smaller tasks: 
+
+1. Implement the persistence of join metadata as attributes into X509
+  certificates.
+2. Implement the WorkloadIdentity resource and CRUD RPCs.
+3. Implement the IssueWorkloadIdentity RPC without templating or rules
+  evaluation.
+4. Implement rules and template evaluation.
+5. Implement label-based IssueWorkloadIdentity RPC functionality.
+6. Implement changes in `tbot` to support using the IssueWorkloadIdentity RPC.
+7. Implement the `tctl workload-identity test` command.
+
+## Alternatives
+
+### Role Templating and `tbot` SPIFFE ID templating
+
+One alternative to introducing a new resource is to introduce support for role
+templating of the SPIFFE fields.
+
+This would still require work to propagate the join metadata into the
+certificate as trait-like things for the purposes of RBAC evaluation.
+
+We'd also need to implement similar templating into the `tbot` configuration
+itself, we could likely reuse much of the same functionality here.
+
+Positives:
+
+1. "Simpler" to implement (but not significantly so)
+2. Keeps configuration within an existing Teleport primitive (roles) rather than
+  further spreading it out into a new resource.
+
+Negatives:
+
+1. TBot configuration still remains important in the SVID issuance process,
+  meaning you still need to manage configuration for a large fleet of `tbot`
+  instances.
+2. The configuration is split between Teleport resources and the `tbot` 
+  configuration. If these drift, SVIDs will fail to be issued.
+3. Teleport Roles are already pretty complex, and this is introducing further
+  complexity.
+
+### CEL vs Teleport Predicate Language
```

Comment:
I think if we're going to support CEL it should be a broader effort that we support in more than one place in the product, i.e. in the K8s rules as well. I'd bring this up in the broader engineering group, and if there's interest in spiking it here great, but I'd hesitate to just move forward with CEL here without that larger buy-in.

Reply:
Agreed - I'll write up a summary and pop it into the engineering channels to collect some thoughts.

Reply:
Discussed with the wider org - consistency seems more valuable than the benefits of CEL at this time.